### DISCUSSION TITLE
Best way to create reusable library for quarkus apps

### DISCUSSION THREAD

**Discussion by ivannov** [quarkusio/quarkus#27587] [Q&A] [answered]
Hi everyone!

I am developing a few microservices with Quarkus. And I'd like them to reuse a bit of functionality including:

* CDI (Arc is also fine)
* MP REST Client
* JSON Binding with Jackson

I don't need that library to be a Quarkus app of its own - it won't be deployed and ran individually, it's just going to be packaged inside other projects.

My microservices don't use GraalVM or any AOT whatsoever.

So what's the best way to implement it? Is it as a quarkus extension? Or would I lose anything if I just implement it as a (Maven) library?

In the latter case, what would be the way to access the Jakarta EE / MicroProfile implementations?

**Comment by gsmet** [comment, marked as answer]
Hi,

Not sure I understand your question: these libraries are for use by Quarkus projects alone or you plan to share them with other non-Quarkus projects?

If it's for Quarkus projects only:
- just make them simple library jars
- just depend on Quarkus extensions (`quarkus-arc`, `quarkus-rest-client` (or the reactive one if you're using RESTEasy Reactive), `quarkus-jackson`)
- you might have to create Jandex indexes/add empty beans.xml in the jars if you have CDI beans there or other things that need to be indexed. See https://stackoverflow.com/a/55513723/5043585 for details

If you plan to share your libraries with non Quarkus projects:
- just make them simple library jars
- depend on the API jars/Jackson directly instead of depending on the Quarkus extensions
- same remark about the indexes

If it doesn't answer your question, it probably requires some clarifications :).

**Reply by ivannov** [reply]
Thanks, Guillaume!

It's the first case (library for Quarkus projects only). I'll follow your suggestion.
